lib/Interpreter/CppInterOp.cpp

```diff
@@ -1027,61 +1056,78 @@ namespace Cpp {
-      if (instantiated)
-        return instantiated;
-
-      // Force the instantiation with template params in case of no args
-      // maybe steer instantiation better with arg set returned from
-      // TemplateProxy?
-      instantiated = InstantiateTemplate(candidate, explicit_types.data(),
-                                          explicit_types.size());
-      if (instantiated)
-        return instantiated;
-
-      // join explicit and arg_types
-      std::vector<TemplateArgInfo> total_arg_set;
-      total_arg_set.reserve(explicit_types.size() + arg_types.size());
-      total_arg_set.insert(total_arg_set.end(), explicit_types.begin(),
-                           explicit_types.end());
-      total_arg_set.insert(total_arg_set.end(), arg_types.begin(),
-                           arg_types.end());
-
-      instantiated = InstantiateTemplate(candidate, total_arg_set.data(),
-                                         total_arg_set.size());
-      if (instantiated)
-        return instantiated;
+    llvm::SmallVector<Expr*> Args;
+    for (auto i : arg_types) {
+      QualType Type = QualType::getFromOpaquePtr(i.m_Type);
+      ExprValueKind ExprKind = ExprValueKind::VK_PRValue;
+      if (Type->isReferenceType())
+        ExprKind = ExprValueKind::VK_LValue;
+      Args.push_back(new (getSema().getASTContext())
+                         OpaqueValueExpr(SourceLocation::getFromRawEncoding(1),
+                                         Type.getNonReferenceType(), ExprKind));
     }
-    return nullptr;
+
+    // Create a list of template arguments.
+    llvm::SmallVector<TemplateArgument> TemplateArgs;
+    TemplateArgs.reserve(explicit_types.size());
+    for (auto explicit_type : explicit_types) {
+      QualType ArgTy = QualType::getFromOpaquePtr(explicit_type.m_Type);
+      if (explicit_type.m_IntegralValue) {
+        // We have a non-type template parameter. Create an integral value from
+        // the string representation.
+        auto Res = llvm::APSInt(explicit_type.m_IntegralValue);
+        Res = Res.extOrTrunc(C.getIntWidth(ArgTy));
+        TemplateArgs.push_back(TemplateArgument(C, Res, ArgTy));
+      } else {
+        TemplateArgs.push_back(ArgTy);
+      }
+    }
+
+    TemplateArgumentListInfo TLI{};
+    for (auto TA : TemplateArgs)
+      TLI.addArgument(
+          S.getTrivialTemplateArgumentLoc(TA, QualType(), SourceLocation()));
+
+    OverloadCandidateSet Overloads(
+        SourceLocation(), OverloadCandidateSet::CandidateSetKind::CSK_Normal);
+    for (void* i : candidates) {
+      Decl* D = static_cast<Decl*>(i);
+      if (auto* MD = dyn_cast<CXXMethodDecl>(D)) {
+        S.AddMethodCandidate(
+            MD, DeclAccessPair::make(MD, MD->getAccess()), MD->getParent(),
+            C.getTypeDeclType(MD->getParent()),
+            Expr::Classification::makeSimpleLValue(), Args, Overloads);
+      } else if (auto* FD = dyn_cast<FunctionDecl>(D)) {
```

Comment:
[AddOverloadCandidate](https://clang.llvm.org/doxygen/classclang_1_1Sema.html#affc2ef4040a1ebccc5e3a59d68807f85) already calls `AddMethodCandidate` if we got a `CXXMethodDecl`...